### ISSUE TITLE
AAPS updates Insight pump basal profile every 5 minutes

### ISSUE THREAD

**Issue by nielsmaerten** [MilosKozak/AndroidAPS#2201]
**AndroidAPS Build:** 2.5.1
**Pump:** Accu-Chek Insight (Spirit Combo probably as well. Haven't tested this yet)

_This issue is related to #1607. I did some digging and managed to consistently reproduce the bug_

If your basal profile has 2 decimals in one of its timeslots, using the 'Profile Switch' feature may result in a profile that (internally) has timeslots with 3 decimals. 
This profile can't be correctly synced to the pump, causing AndroidAPS to try and update it on every connection (every 5 minutes)

Please take a look at this Google Spreadsheet to see the profiles I used: https://www.dropbox.com/scl/fi/xcnd6zk3oi9cz9oaan6gf/AndroidAPS-Bug-Profiles-with-more-than-2-decimals-cause-rounding-errors.gsheet?dl=0&rlkey=nqih1pzbiy4k1tqj4u78j0arn

**Steps to reproduce the bug:**

1. Start by setting Profile A from the spreadsheet as the basal profile. This can either be a Local Profile or an NS Profile. (I used Local Profile, but it shouldn't matter.)
2. Note how 2 of the timeslots in Profile A have 2 decimals (03:00h, and 04:00h). The Insight Pump supports basal values with up to 2 decimals, so this is fine.
3. Now, use the Profile Switch feature and set the percentage to 120%. Internally, this will result in the basal profile on the pump being updated to Profile B (see spreadsheet). Again, this will cause no issues because none of the values have more than 2 decimals. AndroidAPS will update the profile once on the pump, and everything will work normally.
4. Finally, use Profile Switch again, but this time, set the percentage to 130%. Internally, AndroidAPS will try to update the pump profile to Profile C. 

**Expected behavior:**
- The basal profile on the pump is updated once, and things work the same as in Step 3

**Actual behavior:**
- Because the pump does not support values with 3 decimals, this profile (C) cannot be set on the pump. Instead, the timeslots are rounded to 2 decimals. 
  - This causes AndroidAPS to think the profile on the pump is different from the profile it is using internally, and so it updates the pump profile every time it connects. This causes the pump battery to drain quickly, and a 'Pump basal profile updated' notification is shown every 5 minutes.

**AndroidAPS Logs:**
[AndroidAPS.log](https://github.com/MilosKozak/AndroidAPS/files/3843815/AndroidAPS.log)

**Comment by nielsmaerten** [comment]
After looking through the logs, I think maybe AndroidAPS keeps updating the profile because the _isThisProfileSet()_ always returns false? See lines 6249-6250 and 6089-6090 in the logfile.

**Comment by fra-orolo** [comment]
I observed this problem just once with a with a percentage-multiplied profile. After increasing the percentage (by 3 %) the problem went away and was not reproducible anymore.

**Comment by Paulusma** [comment]
I had this issue yesterday setting profile to 130%. Couldn't understand why i got connection timeout issues on the pump. 

Changed %% settings as per suggestion => works. Thx Maarten for the digging!

**Comment by nielsmaerten** [comment]
@Paulusma @fra-orolo 
Probably what happened here is that when you first increased the profile, it had 3 decimals somewhere.
If you then increase by a few more percent, at some point you're gonna hit an internal profile that has 2 decimals in all blocks. Then the rounding error doesn't occur :)

@TebbeUbben @MilosKozak 
I think to fully eliminate this bug, AndroidAPS would have to round all profileblocks to 2 decimals before updating it on the pump. That will prevent this bug from happening down the line.
Please let me know if there's anything I can do to help fix this. I've been exploring the sourcecode myself but admittedly, I'm no expert when it comes to Android or Java

**Comment by Philoul** [comment]
Same issue on my side... I also think it's a rounding problem.
I recently changed my profile with an accuracy of 0.05U whereas previously I had rounded to 0.1 U (I never had any problem before with temp profiles made with multiples of 10%, ie 80%, 90%)

**Comment by wootmasterslick** [comment]
Thanks for finding this. After we adjusted the profile tot the autotune recommendation with 2 decimals we keep getting this problem and couldnt find the cause. Now I know I’ll change the profile to 1decimal. Perhaps @tebbeubben knows how to adapt the code to round it before writing to the pump. I believe he wrote the Insight drivers.

**Comment by Elektrocat** [comment]
Thanks for that - i have the same problem with the insight and now i know why this is happening. (AAPS 2.5.1) Would be nice if it can be fixed in the future because rounding to 0.1 is a bit...hmmm.

**Comment by fra-orolo** [comment]
The right fix to do is, to convert the units [double] to centiUnits [uInt16] in both profiles (Pump and AAPS) in the same way as it is written into the pump (Code for this sits in the setters of ByteBuf). 
Then comparing the resulting integers is stable.  I am currently testing this fix in my build.
The conversions to and from integer in ByteBuf are round-trip stable.


@TebbeUbben 
I'm not quite sure why there are 2 ways of rounding factors (>5u ? 0.1U : 0.01U), is this justified by the pump internals? Does it change the written centiUnit accuracy to deciUnit if the numbers are larger? At least from the BT protocol there is no difference.

**Comment by Philoul** [comment]
I think this bug is fixed since last commit done by Tebbe just before 2.6.0 Release : https://github.com/MilosKozak/AndroidAPS/commit/35096288e2e1e2b974a4bc47afe16d210a802614
(Line 510)...
I checked with Profil set at 99%, 101% and 102% and everything is Ok with last update (2.6.0...)

**Comment by nielsmaerten** [comment]
Can confirm this issue has been resolved as of version 2.6.0. Thanks @TebbeUbben for the fix! 🥇  
I ran the same steps as in the initial bug report and also tried switching the profile to 95%, 99%, 101%, 105%,... it all seems to work fine now.  
Since this particular bug no longer occurs, I'll close this issue.

As for @fra-orolo 's remark: I think the distinction between <5 or >=5 has indeed to do with how the pump handles basal. When setting the basal profile, the max precision is 0.01 when the value is < 5, and 0.1 when it is >=5. Check this video of me updating basal on the pump to see it in action: https://photos.app.goo.gl/K5RY4WGhzLtU3FjJ7
